### ISSUE TITLE
dotnet sln add does not work for .msbuildproj

### ISSUE THREAD

**Issue by bash** [dotnet/sdk#11292]
## Steps to reproduce
1. Create a new solution: `dotnet new sln`
2. Create a new project with the extension `.msbuildproj` (Foo.msbuildproj):
    ```xml
    <?xml version="1.0" encoding="utf-8"?>
    <Project Sdk="Microsoft.Build.NoTargets/1.0.94">
        <PropertyGroup>
            <TargetFramework>netstandard2.1</TargetFramework>
        </PropertyGroup>
    </Project>
    ```
3. Add the project to the solution: `dotnet sln add Foo.msbuildproj`

## Expected behaviour
```
Project `Foo.csproj` added to the solution.
```

## Actual behaviour
```
Project '/.../Foo.msbuildproj' has an unknown project type and cannot be added to the solution file. Contact your SDK provider for support.
```

**Comment by kkirkfield** [comment]
# TLDR / Current Workaround

If you own the custom SDK you are importing and you want to have a custom file extension, then set the `DefaultProjectTypeGuid` property in your Sdk.props file like below.

```xml
<Project>

  <PropertyGroup>
    <DefaultProjectTypeGuid>{2F08BC15-189B-4804-B644-653F34C968A8}</DefaultProjectTypeGuid>
  </PropertyGroup>

</Project>

```

If you don't own the custom SDK and can't make changes to it, then add a `ProjectTypeGuids` property to your project with a random GUID like below. Then your project file extension can be whatever you want.

```xml
<Project Sdk="MyCustom.Sdk">

  <PropertyGroup>
    <ProjectTypeGuids>{2F08BC15-189B-4804-B644-653F34C968A8}</ProjectTypeGuids>
  </PropertyGroup>

</Project>

```

# Research

Currently the dotnet CLI gives an error when trying to add a project to a solution that has an `.msbuildproj` file extension. This file extension is commonly used for projects using the [Microsoft.Build.NoTargets SDK](https://github.com/microsoft/MSBuildSdks/tree/main/src/NoTargets) and other similar custom SDKs.

The error is being set at this location because the project type GUID is null.

https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Cli/dotnet/SlnFileExtensions.cs#L64-L71

The project type GUID is set to null here because the project root element does not have a `ProjectTypeGuids` property set, and there is no `DefaultProjectTypeGuid` property set. `GetProjectTypeGuid()` does not feel like the correct place to define a new value for msbuildproj files. A better option is to set a `DefaultProjectTypeGuid` property for msbuildproj files.

https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Cli/dotnet/SlnFileExtensions.cs#L59

https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Cli/dotnet/ProjectRootElementExtensions.cs#L11-L22

`GetDefaultProjectTypeGuid()` below is where the default value is being received.

https://github.com/dotnet/sdk/blob/d22033010f239b58284a0157a4657c4c56eea30c/src/Cli/dotnet/ProjectInstanceExtensions.cs#L23-L26

Different project SDKs in the dotnet SDK set the `DefaultProjectTypeGuid` property. To avoid projects using the Microsoft.Build.NoTargets SDK and other custom SDKs having to define a project type GUID, we can add a `DefaultProjectTypeGuid` property for projects using the msbuildproj extension. Custom SDKs should still opt to have their own file extension and set the `DefaultProjectTypeGuid` property, but this gives a fallback for msbuildproj files where the SDK doesn't define this.

See the below example of one place where the `DefaultProjectTypeGuid` property is set based on file extension. This specific example is to fix a cross targeting bug, but we can apply the same logic.

https://github.com/dotnet/sdk/blob/c53c8ac81e1bcb8118c31bc4f5622396c72e9959/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.CrossTargeting.targets#L104-L109

Originally I was going to do a pull request with the fix for this after line 42 below. That would allow it to fallback after the language targets and the cross targeting target referenced above would run.

https://github.com/dotnet/sdk/blob/172b214731f8669f8cfc3c20816da14b7f6c8d9f/src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets#L37-L42

While this did work for project files using a dotnet SDK, it didn't work for project files targeting custom SDKs like Microsoft.Build.NoTargets. For the fallback to apply to all projects regardless of SDK, I believe this would need to be added somewhere in [MSBuild](https://github.com/dotnet/msbuild). I'm currently having trouble finding the best place to put this in MSBuild as the prop and target imports are hard to follow. If anyone has any ideas let me know below.

# See also
- https://github.com/microsoft/MSBuildSdks/issues/155
- https://github.com/dotnet/msbuild/issues/3558

**Comment by kkirkfield** [comment]
@KalleOlaviNiemitalo I looked through the DotNet SDK and MSBuild and didn't see anywhere that uses these project type GUIDs other than for the solution file, so I don't think it matters what GUID you choose as long as it is unique. I think Visual Studio may use these GUIDs in some way for example to enable/disable tooling for different workloads, but I can't be sure of that.

**Comment by kkirkfield** [comment]
@KalleOlaviNiemitalo  If you use the Microsoft.Build.NoTargets SDK and need multitargeting, and you want to be able to open and build in Visual Studio, then the GUID you choose may be important. See this comment and thread for more info: https://github.com/microsoft/MSBuildSdks/issues/155#issuecomment-1341764985

**Comment by ceztko** [comment]
Any update on this?
